include/mata/nfa/delta.hh

```diff
@@ -86,6 +98,64 @@ public:
     using super::find;
     iterator find(const Symbol symbol) { return super::find({ symbol, {} }); }
     const_iterator find(const Symbol symbol) const { return super::find({ symbol, {} }); }
+
+    /**
+     * Iterator over moves. It iterates over pairs (symbol, target state) for a current source state whose state post
+     *  we iterate.
+     */
+    struct moves_const_iterator {
+    private:
+        const std::vector<SymbolPost>& m_symbol_posts{};
+        std::vector<SymbolPost>::const_iterator m_symbol_post_it{};
+        StateSet::const_iterator m_target_states_it{};
+        bool m_is_end{ false };
+        Move m_move{};
+
+    public:
+        using iterator_category = std::bidirectional_iterator_tag;
+        using value_type = int;
+        using difference_type = int;
+        using pointer = int*;
+        using reference = int&;
+
+        explicit moves_const_iterator(const std::vector<SymbolPost>& symbol_posts, bool is_end = false);
+
+        moves_const_iterator(const std::vector<SymbolPost>& symbol_posts,
+                             std::vector<SymbolPost>::const_iterator symbol_posts_it,
+                             StateSet::const_iterator target_states_it, bool is_end = false);
+
+        moves_const_iterator(const moves_const_iterator& other) = default;
+
+        const Move& operator*() const { return m_move; }
+
+        // Prefix increment
+        moves_const_iterator& operator++();
+        // Postfix increment
+        const moves_const_iterator operator++(int);
+
+        moves_const_iterator& operator=(const moves_const_iterator& x);
+
+        bool operator==(const moves_const_iterator& other) const;
+        bool operator!=(const moves_const_iterator& other) const { return !(*this == other); };
+    };
+
+    moves_const_iterator moves_cbegin() const { return moves_const_iterator(ToVector()); }
+    moves_const_iterator moves_cend() const { return moves_const_iterator(ToVector(), true); }
+    moves_const_iterator moves_begin() const { return moves_cbegin(); }
+    moves_const_iterator moves_end() const { return moves_cend(); }
+
+    struct MovesIterator {
```

Comment:
We need a struct to have `begin()` and `end()` functions and returning the iterator `moves_const_iterator`. This is the idiomatic way to implement this functionality so that you can do range for loops.

Reply:
It could be named 'Moves'. Do you find it better? As in, you would get `Moves::begin()` and `Moves::end()` to iterate through moves for a given state.

Reply:
yes yes, Moves better